tests/core/actions/test_action_run_slot_rejections.py

```diff
@@ -0,0 +1,421 @@
+            "step_id": "ask_payment_recipient",
+            "frame_type": "regular",
+            "type": "flow",
+        },
+        {
+            "frame_id": "6Z7PSTRM",
+            "flow_id": "pattern_ask_collect_information",
+            "step_id": "start",
+            "collect_information": "payment_recipient",
+            "rejections": [],
+            "type": "pattern_collect_information",
+        },
+    ]
+
+    tracker = DialogueStateTracker.from_events(
+        sender_id=uuid.uuid4().hex,
+        evts=[
+            UserUttered("I want to make a payment."),
+            SlotSet("dialogue_stack", dialogue_stack),
+        ],
+        slots=[
+            TextSlot("payment_recipient", mappings=[]),
+            AnySlot("dialogue_stack", mappings=[]),
+        ],
+    )
+
+    action_run_slot_rejections = ActionRunSlotRejections()
+    events = await action_run_slot_rejections.run(
+        output_channel=default_channel,
+        nlg=rejection_test_nlg,
+        tracker=tracker,
+        domain=rejection_test_domain,
+    )
+
+    assert events == []
+
+
+async def test_action_run_slot_rejections_top_frame_slot_not_been_set(
+    default_channel: OutputChannel,
+    rejection_test_nlg: TemplatedNaturalLanguageGenerator,
+    rejection_test_domain: Domain,
+    capsys: CaptureFixture,
+) -> None:
+    dialogue_stack = [
+        {
+            "frame_id": "4YL3KDBR",
+            "flow_id": "setup_recurrent_payment",
+            "step_id": "ask_payment_type",
+            "frame_type": "regular",
+            "type": "flow",
+        },
+        {
+            "frame_id": "6Z7PSTRM",
+            "flow_id": "pattern_ask_collect_information",
+            "step_id": "start",
+            "collect_information": "recurrent_payment_type",
+            "rejections": [
+                {
+                    "if": 'not ({"direct debit" "standing order"} contains recurrent_payment_type)',  # noqa: E501
+                    "utter": "utter_invalid_recurrent_payment_type",
+                }
+            ],
+            "type": "pattern_collect_information",
+        },
+    ]
+
+    tracker = DialogueStateTracker.from_events(
+        sender_id=uuid.uuid4().hex,
+        evts=[
+            UserUttered("i want to setup a new recurrent payment."),
+            SlotSet("dialogue_stack", dialogue_stack),
+        ],
+        slots=[
+            TextSlot("recurrent_payment_type", mappings=[]),
+            AnySlot("dialogue_stack", mappings=[]),
+        ],
+    )
+
+    action_run_slot_rejections = ActionRunSlotRejections()
+    events = await action_run_slot_rejections.run(
+        output_channel=default_channel,
+        nlg=rejection_test_nlg,
+        tracker=tracker,
+        domain=rejection_test_domain,
+    )
+
+    assert events == []
+    out = capsys.readouterr().out
+    assert (
+        "[debug    ] first.collect.slot.not.set     slot_name=recurrent_payment_type slot_value=None"  # noqa: E501
+        in out
+    )
```

Comment:
Hmhmhm, testing for debug output might be going a bit too far. With it you are testing here for specific internal workings and not for externally relevant behavior (unless we say we specifically want to guarantee that this specific debug output is always there). The issue is that the test case becomes more prone to fail even for valid changes to the code.

I would say testing for an error output being there is okay, as you do below, because it's important for the user.

Up to you I would say, but I wanted to raise the point :)

Reply:
Gotcha, I added these explicit assertions for the log message in order to signpost which code branch I'm testing (since quite a few branches are returning empty list of events). I will update so it matches the approaches I took in the other tests.